src/components/editor/MonacoEditor.tsx

```diff
@@ -22,35 +29,128 @@ function MonacoEditor({ disabled, value, path, height, onChange }: Props) {
         null
     );
     const heightVal = height ?? DEFAULT_HEIGHT;
+    const [isChanging, setIsChanging] = useState<boolean | null>(null);
+    const [hasServerChanges, setHasServerChanges] = useState<any | null>(null);
+    const [showServerDiff, setShowServerDiff] = useState(false);
 
-    const handlers = {
-        change: debounce(async (val) => {
-            if (editorRef.current && onChange) {
-                onChange(JSON.parse(val), path);
+    const serverUpdate = useZustandStore<
+        EditorStoreState<DraftSpecQuery>,
+        EditorStoreState<DraftSpecQuery>['serverUpdate']
+    >((state) => state.serverUpdate);
+
+    useEffect(() => {
+        if (editorRef.current) {
+            const currentStringValue = editorRef.current.getValue();
+
+            if (currentStringValue) {
+                const currentEditorValue = JSON.parse(currentStringValue);
+
+                if (!isEqual(currentEditorValue, serverUpdate)) {
+                    setHasServerChanges(serverUpdate);
+                } else {
+                    setHasServerChanges(null);
+                }
+            }
+        }
+
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [serverUpdate]);
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    const bar = useCallback(
+        debounce((val: string | undefined) => {
+            console.log('debounce', {
+                val,
+                curr: editorRef.current?.getValue(),
+            });
```

Comment:
Leaving in some logging as we'll need to test the crap out of all this debouncing and stuff. Might make it easier if another dev runs into issues.